src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -1125,78 +1138,108 @@ public override Stream EndGetRequestStream(IAsyncResult asyncResult)
-                        request.Content ??= new ByteArrayContent(Array.Empty<byte>());
+            _sendRequestMessage.Version = ProtocolVersion;
 
-                        request.Content.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
-                    }
-                    else
-                    {
-                        request.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
-                    }
-                }
+            _sendRequestMessage.Headers.TransferEncodingChunked = SendChunked;
 
-                request.Headers.TransferEncodingChunked = SendChunked;
+            if (KeepAlive)
+            {
+                _sendRequestMessage.Headers.Connection.Add(HttpKnownHeaderNames.KeepAlive);
+            }
+            else
+            {
+                _sendRequestMessage.Headers.ConnectionClose = true;
+            }
 
-                if (KeepAlive)
-                {
-                    request.Headers.Connection.Add(HttpKnownHeaderNames.KeepAlive);
-                }
-                else
-                {
-                    request.Headers.ConnectionClose = true;
-                }
+            _sendRequestMessage.Version = ProtocolVersion;
+            HttpCompletionOption completionOption = _allowReadStreamBuffering ?
+                            HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead;
+            // If we're not buffering, there is no way to open the connection and not send the request without async.
+            // So we should use Async, if we're not buffering.
+            _sendRequestTask = async || !AllowWriteStreamBuffering ?
+                _httpClient.SendAsync(_sendRequestMessage, completionOption, _sendRequestCts.Token) :
+                Task.FromResult(_httpClient.Send(_sendRequestMessage, completionOption, _sendRequestCts.Token));
+
+            return _sendRequestTask!;
+        }
 
-                if (_servicePoint?.Expect100Continue == true)
+        private async Task<WebResponse> HandleResponse(bool async)
+        {
+            try
+            {
+                _requestStream?.EndWriteOnStreamBuffer();
+            }
+            catch (ObjectDisposedException)
+            {
+                // If the request stream was disposed, don't do anything.
+            }
+
+            HttpContent content;
+            if (_requestStream is not null)
+            {
+                content = new StreamContent(new HttpClientContentStream(_requestStream.GetBuffer()));
+                if (AllowWriteStreamBuffering)
                 {
-                    request.Headers.ExpectContinue = true;
+                    content.Headers.ContentLength = _requestStream.GetBuffer().ReadBytesAvailable;
                 }
+            }
+            else
+            {
+                content = new ByteArrayContent(Array.Empty<byte>());
```

Comment:
Nit: This could remain as it was before the PR where we only lazily allocate it when we have to for headers. Unless I'm missing some detail?